grails-app/migrations/views/cycle-count-session.sql

```diff
@@ -1,40 +1,74 @@
-	FROM product_availability
-	JOIN location on product_availability.location_id = location.id
-	JOIN product on product_availability.product_id = product.id
-	JOIN inventory on location.inventory_id = inventory.id
-    LEFT OUTER JOIN inventory_level on product.id = inventory_level.product_id AND inventory.id = inventory_level.inventory_id
-    LEFT OUTER JOIN cycle_count_request ON product.id = cycle_count_request.product_id AND location.id = cycle_count_request.facility_id
-    LEFT OUTER JOIN cycle_count ON cycle_count_request.cycle_count_id = cycle_count.id
-    WHERE product_availability.quantity_on_hand > 0
-       AND (cycle_count_request.id IS NULL OR (cycle_count_request.status <> 'COMPLETED' AND cycle_count_request.status <> 'CANCELED'))
-	GROUP BY location.id, product.id, abc_class, cycle_count_request.id
-    ORDER BY abc_class, inventory_item_count desc
-);
+CREATE OR REPLACE VIEW cycle_count_session AS
+(
+SELECT
+    # ID hash of product.id and location.id
+    CRC32(CONCAT(location.id, product.id))                                       as id,
+    location.inventory_id                                                                 as inventory_id,
+    product.id                                                                   as product_id,
+    location.id                                                                  as facility_id,
+
+    -- Cycle count request ID (deprecated)
+    cycle_count_request_summary.cycle_count_request_id as cycle_count_request_id,
+
+    -- ABC Classification
+    COALESCE(inventory_level_summary.abc_class, product.abc_class)               as abc_class,
+
+    -- Status of any pending cycle counts
+    -- Consolidate the statuses down to a single field representing the status of the candidate itself.
+    cycle_count_request_summary.status as status,
+
+    # Inventory Item Count
+    count(product_availability.id)                                               as inventory_item_count,
+
+    # Comma-separated list of internal locations
+    GROUP_CONCAT(DISTINCT product_availability.bin_location_name)                as internal_locations,
+
+    # Quantities by SKU
+    sum(product_availability.quantity_on_hand)                                   as quantity_on_hand,
+    sum(product_availability.quantity_available_to_promise)                      as quantity_available,
+
+    # Stuff we need to include now or wait for the materialized view
+    -- Negative item count
+    (SELECT SUM(CASE WHEN pai.quantity_on_hand < 0 THEN 1 ELSE 0 END)
+     FROM product_availability pai
+     WHERE pai.product_id = product_availability.product_id
+       AND pai.location_id = product_availability.location_id)                   as negative_item_count,
+
+    -- Date last counted
+    (select date_counted
+     from product_count_history
+     where product_count_history.product_id = product_availability.product_id
+       and product_count_history.inventory_id = location.inventory_id) as date_last_count,
```

Comment:
See my comment in product_count_history. Does this not have the potentialy to return multiple rows for cycle count?

Reply:
Good call. I think I intended for there to be GROUP BYs either in the lower view definitions or this subquery. But the group by got lost somewhere along the way.

Reply:
Added group by to product_count_history